### PR TITLE
Add code for a first version of a functioning FFI library

### PR DESCRIPTION
**Pull request by nesium**
This PR started as a proof-of-concept but a few patterns emerged that seem like a good way forward.

The current version of the library is able to:
- Connect to a server
- Send and receive messages and report received messages to an `AccountObserver`
- Request a roster and forward it to the `AccountObserver`

Let's look at the areas where I didn't spend any time on and why:
- Naming methods in the FFI library. This is not important at this time because I'll add another wrapper in the app which hides the FFI libraries details and exposes a proper Swift-like interface. So we'll be flexible with what happens between the wrapper and the core library.
- Error handling. Nested error enums don't seem to go through UniFFI. Since there are multiple different error types at each layer I haven't taken the time to find a way to flatten them. This will be important in the next step though so that we have an idea what's going on when something doesn't work in the app.
- Using proper `id`s when sending `iq`s to the server and matching them with the server responses. If you don't want to take over the code I'll probably make a list of which features are high on our priority list then do research how the XMPP protocol looks like to support these features and find a proper abstraction.
- Logging. I've tried an `os_log` Rust wrapper briefl,y conditionally compiled for Apple targets, but couldn't get it to work. I've left the plain logger in there but made it optional to enable it so that we don't have too much noise in the app if we don't want to. Mid-term I'd like either handle the app the logging (so the logger would forward its messages over FFI) or go straight to `os_log` which can be filtered in the console.

I had a crash when registering an `iq` and `message` handler for some reason. Might be a problem with my understanding of Rust and putting them in a `Box` or a limitation of the library (probably not?). This is up for further investigation.

What's good:
- The `TryFrom` trait seems like a good way to convert Stanzas into Rust structs
- The channels seem like a good thread synchronization mechanism for this use-case.

### CODE REVIEW

File: prose_core_client_ffi/src/account_observer.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+use crate::Roster;
+
+use super::types::message::Message;
+
+#[allow(non_snake_case)]
+pub trait AccountObserver: Send + Sync {
+    fn didConnect(&self);
```
Comment (valeriansaliou) [inline prose_core_client_ffi/src/account_observer.rs:7]: Is the use of camel case done on purpose here? Eg. exposing a camel-case FFI API? As Rust coding patterns / community conventions enforce snake-case.

**Comment by valeriansaliou** [comment]
Thanks Marc. LGTM. Just posted a comment regarding the camel-case thing, then we're ready to merge. 👍🏻

File: prose_core_client_ffi/src/account_observer.rs (hunk 1 above)
Comment (nesium) [reply prose_core_client_ffi/src/account_observer.rs:7]: Yes, the `AccountObserver` is a trait that needs to be implemented by a consumer of the FFI library. We'd either force snake-case everywhere in the FFI or stick to camel case. What's your preference?

Comment (valeriansaliou) [reply prose_core_client_ffi/src/account_observer.rs:7]: If I get this right, the `uniffi` crate would then map those declared Rust names to eg. Swift names for the referred-to function. If that's so, then we should keep it camel in the Rust code yes, unless `uniffi` provides a way to "camelify" snake-case names? That way we can stick to pure-Rust conventions in our Rust code.

**Comment by nesium** [comment]
> Also, since there's a lot of logic in the FFI library now, could this be ported into the main library?

Absolutely, but I did not for two reasons:
- I didn't know how we're working together on the library and which parts you'd want to take. So my priority was to get all the parts going first to have an understanding if UniFFI works for us and that the communication between Rust and Swift works as expected.
- I didn't feel comfortable to put the code in the structure you've created. I'm not 100% sure if that would even be the right way to do it, or if it would be better to let the structure emerge from the patterns that prove useful. I'm leaning towards the latter. The other thing is that even if I wanted my Rust skills wouldn't let me since having this many objects communicating with each other is an ownership horror for a Rust rookie. I wanted to spend the time to be productive and achieve my first goal instead of fighting the compiler.

File: prose_core_client_ffi/src/account_observer.rs (hunk 1 above)
Comment (nesium) [reply prose_core_client_ffi/src/account_observer.rs:7]: That's how the corresponding Swift code looks like…

```swift
extension ViewModel: AccountObserver {
  func didConnect() {
    print("Did connect")
  }

  func didDisconnect() {
    print("Did disconnect")
  }

  func didReceiveMessage(message: ProseCoreClientFFI.Message) {
    DispatchQueue.main.async {
      self.messages.append(Message(from: message.from, body: message.body))
    }
  }

  func didReceiveRoster(roster: Roster) {
    print(roster)
  }
}
```

**Comment by valeriansaliou** [comment]
I get you. My code is probably too Rusty as I've already worked of Rust projects in the past. Probably too abstract, which causes me some headaches as well integrating w/ `libstrophe`. I agree on your plan then, once you settled on a good working prototype structure, we can then work together to re-integrate the logics you've built into the main library, and organize it in a Rusty way (w/ a bit more abstractions, generics and the like).

File: prose_core_client_ffi/src/account_observer.rs (hunk 1 above)
Comment (nesium) [reply prose_core_client_ffi/src/account_observer.rs:7]: I'm not sure that converting the case-style on the fly would even work. I haven't looked to deep in the innards of the UniFFI implementation but I'd imaging that the function calls will still have to access the good old exported C symbols. At least for the Swift output there is [no such option](https://mozilla.github.io/uniffi-rs/swift/configuration.html).

Comment (valeriansaliou) [reply prose_core_client_ffi/src/account_observer.rs:7]: Let's stick to camel-case in the Rust code then 🙂

**Comment by nesium** [comment]
I mean if you want to you can make it your own now, as long as you leave the FFI interface as-is. My plan was to do a little bit of clean-up in the sidebar of the app. Then weave the library in so that the roster is displayed there and we can send and receive messages. If you can whip it into shape during that time, by all means.

So far I haven't thought about where to go from there though. You could also do some prioritization and planning. Point out the XEPs we need to support for the next features. What do you think should be next for us?

**Pull request merged**